rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/SortedUniqueSet.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * ProActive Parallel Suite(TM):
+ * The Open Source library for parallel and distributed
+ * Workflows & Scheduling, Orchestration, Cloud Automation
+ * and Big Data Analysis on Enterprise Grids & Clouds.
+ *
+ * Copyright (c) 2007 - 2017 ActiveEon
+ * Contact: contact@activeeon.com
+ *
+ * This library is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Affero General Public License
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.resourcemanager.common.event;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+
+interface Unique {
+
+    /**
+     * @return key which should be unique
+     */
+    String getKey();
+
+    /**
+     * @return counter which is used to keep sorted collection
+     */
+    long getCounter();
+}
+
+/**
+ * Set based on #getKey() that provides log(N) to add and remove, but keeps sorted version of items based on #getCounter().
+ * @param <T>
+ */
+public class SortedUniqueSet<T extends Unique & Comparable<T> & Serializable> implements Serializable {
```

Comment:
Can't we use a java.util.TreeMap in place of the SortedUniqueSet?

Reply:
I tried to use, but then I found that I am not obeying "equals compareTo contract".
Basically, I want sorted set which is unique by 'key' and sorted by 'counter'.